### PR TITLE
Proper™ input handling

### PR DESCRIPTION
**Pull request by NegativeScript**
Excuse the branch name. I tried.

### CODE REVIEW

File: furusiya/main_interface.py [hunk 1]
```diff
@@ -33,13 +37,22 @@ def run(cls, to_run):
             panel=(SCREEN_WIDTH, PANEL_HEIGHT),
             fps_limit=LIMIT_FPS
         )
+        cls.saveload = SaveLoad(cls)
+
+        from model.key_binder import KeyBinder
```
Comment (nightblade9) [inline furusiya/main_interface.py]: Imports should always be at the top of the file.

File: furusiya/main_interface.py [hunk 2]
```diff
@@ -33,13 +37,22 @@ def run(cls, to_run):
             panel=(SCREEN_WIDTH, PANEL_HEIGHT),
             fps_limit=LIMIT_FPS
         )
+        cls.saveload = SaveLoad(cls)
```
Comment (nightblade9) [inline furusiya/main_interface.py]: This makes my brain hurt. Why isn't this `Game.saveload`?

File: furusiya/main_interface.py [hunk 3]
```diff
@@ -24,6 +25,9 @@ class Game:
     current_turn = None
     playing = False  # True when in-game, false otherwise
 
+    saveload = None
```
Comment (nightblade9) [inline furusiya/main_interface.py]: Needs a better name. Should really be `save_load` according to the class name.

File: furusiya/main_interface.py [hunk 4]
```diff
@@ -24,6 +25,9 @@ class Game:
...
+    keybinder = None
```
Comment (nightblade9) [inline furusiya/main_interface.py]: Should be `key_binder` which is a weird name. Binder (thing that holds papers) with a keys in it?

File: furusiya/main_interface.py [hunk 5]
```diff
@@ -33,13 +37,22 @@ def run(cls, to_run):
...
+        cls.keybinder = KeyBinder(cls)
 
         to_run()
 
         print("Terminating ...")
 
         file_watcher.stop()
 
+    @classmethod
+    def set_keybinds_and_events(cls):
```
Comment (nightblade9) [inline furusiya/main_interface.py]: Again, why not just set these on `Game`? Why pass in `cls`? To me, this should be a static method that operates on `Game.keybinder`.

File: furusiya/main.py [hunk 6]
```diff
@@ -6,16 +6,65 @@
...
+from model.entities.party.stallion import Stallion
+from model.maps import generators
+from model.maps.area_map import AreaMap
+from model.maps.generators import DungeonGenerator
+from main_interface import Game, message
+from constants import SCREEN_WIDTH, SCREEN_HEIGHT, MAP_WIDTH, MAP_HEIGHT
 import random
 
 
+def new_game():
```
Comment (nightblade9) [inline furusiya/main.py]: Needs tests: at least that:
-  `Game.area_map` is initialized with a non-zero size
- `Game.player` is not `None`
- ditto for stallion (you can also check the coordinates are close to the player)
- player coordinates are random and walkable (call `new_game` twice and he should move)
- Inventory is not `None`, it's an empty list
- game state is `playing`
- Game messages are empty

File: furusiya/main.py [hunk 7]
```diff
@@ -6,16 +6,65 @@
...
+    # TODO: what if we spawned in a wall? :/
+    Game.stallion.x = Game.player.x + 1
+    Game.stallion.y = Game.player.y + 1
+    Game.area_map.entities.append(Game.stallion)
+
+    Game.game_state = 'playing'
+    Game.inventory = []
+
+    # create the list of game messages and their colors, starts empty
+    Game.game_msgs = []
```
Comment (nightblade9) [inline furusiya/main.py]: No abbreviations. `game_messages` please.

File: furusiya/view/data/saveload.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import shelve
+
+
+class SaveLoad:
```
Comment (nightblade9) [inline furusiya/view/data/saveload.py]: This shouldn't be in the `view` directory. It should be under some sort of `i/o` type of directory (but not `io` as we've seen with package/namespace shadowing).

File: furusiya/model/key_binder.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
+from model.keys.events import update_callback, quit_event, mousemotion_event
+from model.keys.key_callbacks import *
+
+
+keybinds = {
```
Comment (nightblade9) [inline furusiya/model/key_binder.py]: Shouldn't this be a constant? We're not changing it at runtime.

File: furusiya/model/key_binder.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    'd': inventory_drop,
+
+    'f': bow_callback,
+    'm': mount_callback,
+
+    'r': rest_callback,
+    'R': continuous_rest_callback
+}
+
+events = {
```
Comment (nightblade9) [inline furusiya/model/key_binder.py]: Ditto here.

File: furusiya/model/key_binder.py (hunk 9 above)
Comment (nightblade9) [reply furusiya/model/key_binder.py]: Should also be called `key_bindings`.

File: furusiya/model/key_binder.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+class KeyBinder:
+    def __init__(self, game):
+        self.game = game
+
+    # Keybinds
+    def register_all_keybinds(self):
+        for key, callback in keybinds.items():
+            self.register_keybind(key, callback)
+
+    def suspend_all_keybinds(self):
```
Comment (nightblade9) [inline furusiya/model/key_binder.py]: Is this code even used?

File: furusiya/model/key_binder.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        delattr(self.game.ui.app, f'key_{key}')
+
+    # Events
+    def register_all_events(self):
+        for key, callback in events.items():
+            self.register_event(key, callback)
+
+        self.register_update()
+
+    def suspend_all_events(self):
```
Comment (nightblade9) [inline furusiya/model/key_binder.py]: Do we need this?

File: furusiya/model/key_binder.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    'MOUSEMOTION': mousemotion_event,
+    'MOUSEDOWN': None,
+    'MOUSEUP': None,
+
+    'KEYDOWN': None,
+    'KEYUP': None,
+}
+
+
+class KeyBinder:
```
Comment (nightblade9) [inline furusiya/model/key_binder.py]: Can we get rid of code that we're not using and probably won't need? Or is all of this new code in-use?

File: furusiya/model/keys/events.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+from main_interface import Game
+from model.keys.key_callbacks import exit_to_main_menu_callback
+from model.systems.ai_system import AISystem
+
```
Comment (nightblade9) [inline furusiya/model/keys/events.py]: "Events" makes me/people think event-driven design, which is not what this is; `callbacks` might be a better name.

File: furusiya/model/keys/key_callbacks.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+def pickup_callback(event):
+    for obj in Game.area_map.entities:  # look for an item in the player's tile
+        obj_item = obj.get_component(Item)
+        if (obj.x, obj.y) == (Game.player.x, Game.player.y) and obj_item:
+            obj_item.pick_up()
+            break
+
+
+# Inventory use
+@in_game(pass_turn=False)
```
Comment (nightblade9) [inline furusiya/model/keys/key_callbacks.py]: I really like the clarity of this decorator.

File: furusiya/view/adapter/extensible_app.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
+import tdl
+import time
+
+from tdl.event import get
+
+
+class ExtensibleApp:
+    """
+    Copypasta from tdl.event.App but without __slots__ defined.
+    This hurt to make.
```
Comment (nightblade9) [inline furusiya/view/adapter/extensible_app.py]: +1

File: furusiya/main_interface.py (hunk 1 above)
Comment (NegativeScript) [reply furusiya/main_interface.py]: Sadly, having this at the top of the file resulted in cyclic imports.
I'll have another go at rearranging things.

File: furusiya/main_interface.py (hunk 2 above)
Comment (NegativeScript) [reply furusiya/main_interface.py]: Because `cls` in this case refers to `Game`? Both do just as well.

File: furusiya/main_interface.py (hunk 4 above)
Comment (NegativeScript) [reply furusiya/main_interface.py]: I say this is good as-is. Think keybinds. Keybinder. Makes sense, no?

File: furusiya/model/key_binder.py (hunk 12 above)
Comment (NegativeScript) [reply furusiya/model/key_binder.py]: Probably not. I just wrote the entire fleshed out thing and then integrated the keybinds where appropriate.

I'll remove unused ones.

File: furusiya/main_interface.py (hunk 1 above)
Comment (nightblade9) [reply furusiya/main_interface.py]: That sucks. I never figured out how to get around that. No worries, you can leave it as-is, but please add a comment stating why it's here (cyclical imports)

File: furusiya/main_interface.py (hunk 5 above)
Comment (nightblade9) [reply furusiya/main_interface.py]: I would argue that this is actually less clear than operating directly on `Game`. But sure, whatever works for you.

Comment (nightblade9) [reply furusiya/main_interface.py]: I checked my course notes, and this is apparently the way to go.

File: furusiya/game.py [hunk 17]
```diff
@@ -0,0 +1,19 @@
+class Game:
+    inventory = []
+    draw_bowsight = None
+    player = None
+    stallion = None
+    mouse_coord = (0, 0)
+    auto_target = None
+    target = None
+    game_msgs = []
```
Comment (nightblade9) [inline furusiya/game.py]: Please change this abbreviation as requested elsewhere.

File: furusiya/main.py [hunk 18]
```diff
@@ -1,20 +1,84 @@
...
+from view.map_renderer import MapRenderer
 
-from function_mess import new_game, play_game, load_game
-from main_interface import Game
-from constants import SCREEN_WIDTH, SCREEN_HEIGHT
-import random
+
+def new_game():
+
+    Game.area_map = AreaMap(MAP_WIDTH, MAP_HEIGHT)
```
Comment (nightblade9) [inline furusiya/main.py]: Please add tests for all this logic as mentioned elsewhere. Default values, things not being `None`, etc. is fine -- don't need to test that the return value from other calls is sane.

File: furusiya/game.py (hunk 17 above)
Comment (nightblade9) [inline furusiya/game.py]: Please change this abbreviation as requested elsewhere.

File: furusiya/model/helper_functions/death_functions.py [hunk 19]
```diff
@@ -1,34 +1,34 @@
...
-    player.char = '%'
+import colors
+from model.helper_functions.message import message
+from game import Game
+from model.systems.ai_system import AISystem
+from model.systems.fighter_system import FighterSystem
+from model.systems.xp_system import XPSystem
+
+
+def monster_death(monster):
```
Comment (nightblade9) [inline furusiya/model/helper_functions/death_functions.py]: Please add tests. Important things to test:
- Character changes to `%`
- `blocks` becomes false
- Player gets some non-zero amount of XP
- Monster AI becomes `None`
- `hostile` becomes false

I would write these in a single test, something like `test_monster_death_makes_monster_immobile`

File: furusiya/model/helper_functions/death_functions.py [hunk 20]
```diff
@@ -1,34 +1,34 @@
...
+    FighterSystem.remove_fighter(monster)
+    AISystem.remove_ai(monster)
+
+    monster.original_ai = None
+    monster.name = "{} remains".format(monster.name)
+    monster.hostile = False
+    monster.send_to_back()
+
+
+def player_death(player):
```
Comment (nightblade9) [inline furusiya/model/helper_functions/death_functions.py]: Please add a test; assert that it affects the game state.

File: furusiya/model/keys/util.py [hunk 21]
```diff
@@ -0,0 +1,20 @@
+from game import Game
```
Comment (nightblade9) [inline furusiya/model/keys/util.py]: Please rename this file to `in_game_decorator`

File: furusiya/model/keys/util.py [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+
+
+def in_game(callback=None, *, pass_turn=False):
+    """
+    Decorator for ensuring in-game access only
+    Also marks the current turn as passed.
+    """
+    def decorator(callback):
+        def _inner_function(event):
+            if Game.game_state == 'playing' and Game.current_turn is Game.player:
```
Comment (nightblade9) [inline furusiya/model/keys/util.py]: Please write tests for this. Basically, assert that the `Game.current_turn` value becomes none (or not) if the state is playing etc.

File: furusiya/model/maps/area_map.py [hunk 23]
```diff
@@ -50,7 +50,15 @@ def place_on_random_ground(self, entity):
         entity.x = x
         entity.y = y
         self.entities.append(entity)
-    
+
+    def get_blocking_object_at(self, x, y):
```
Comment (nightblade9) [inline furusiya/model/maps/area_map.py]: Write tests please.

File: furusiya/view/adapter/tdl_adapter.py [hunk 24]
```diff
@@ -17,12 +18,25 @@ def __init__(self, window_title, screen, map, panel, fps_limit=20):
         self.con = tdl.Console(*map)
         self.panel = tdl.Console(*panel)
 
+        self.app = ExtensibleApp()
+
+    @property
+    def event_queue(self):
+        """Private, I know. Not my fault tdl sucks."""
```
Comment (nightblade9) [inline furusiya/view/adapter/tdl_adapter.py]: TDL does not suck. Them's fightin' words!

File: furusiya/view/data/saveload.py [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self, game):
+        self.game = game
+
+    def save(self):
```
Comment (nightblade9) [inline furusiya/view/data/saveload.py]: This needs tests. Maybe something as simple as calling `load(save(x))` and verify that all the properties are still set. We'll need this safety net later when we add stuff and have to worry about how it'll work with existing saves.

**Pull request merged**